File: singleton-pattern/singleton.ts
class Singleton {
  private static instance: Singleton;

  // Making the constructor private so that it can't be instantiated directly
  private constructor() {}

  public static getInstance() {
    if (!Singleton.instance) {
      Singleton.instance = new Singleton();
    }

    return Singleton.instance;
  }
}

// Dry run
const s1 = Singleton.getInstance();
const s2 = Singleton.getInstance();

const bothInstancesSame = s1 === s2;

console.log(`\n Both singleton instances are same: ${bothInstancesSame} \n\n`);

File: singleton-pattern/singleton.py
class Singleton:
    _instance = None

    def __new__(cls, *args, **kwargs):
        if cls._instance is None:
            cls._instance = super(Singleton, cls).__new__(cls)

        return cls._instance


# Using singleton
s1 = Singleton()
s2 = Singleton()

print(f"Both instances are same {s1 is s2}")
